### ISSUE TITLE
approaches to measuring/improving performance?

### ISSUE THREAD

**Issue by rahulbot** [mediacloud/story-indexer#277]
We have seen slow search performance and that hinders internal and external usage significantly. We need to design multiple approaches to measuring search performance and take steps to improve it.

I imagine the first step would be developing some set of "canonical" searches that are a reasonable span of the space of our most common type of searches. With those in hand, we would then need some way to characterize their end-to-end performance in the system (i.e. user-facing speed). If we understood the slowest part of that then we could take steps to speed things up and test again.

Thoughts on how to approach this?

**Comment by philbudne** [comment]
The tool I'd like to be able to open up would be, for every (search)
submit on the web site: a screen with time on the horizontal axis,
with horizontal bars representing time spent in a component (bars may
overlap for requests executing in parallel), like those you can get
from a browser's console window showing a page load, except that any
bar can be "opened" (clicking to rotate an arrow from > to v) to
expand analysis of that transaction, allowing tracing the full page
load from end to end through components (Browswer to Django to News
Search API to ES and back) and perhaps even into the code within those
components.

Essentially, each trace is a tree with expandable/collapsible leaves.

As a tool builder, I imagined that each request includes a list of
guids (which could be as simple as local hostname plus local
timestamp) and each step along the way, the passage of the request or
response is reported to a server via an HTTP post.  Initially the list
of guids is created by the browser: any place along the line a
parallel activity is "forked", a new (hostname, timestamp) pair is
appended to the guid list.  Of course, the hard part is the display.

A few google searches got me to the term "distributed tracing", and
after my initial impulse to add "open source" to the search terms, I
remembered Rahul's preference to use paid services rather than
maintain yet another OSS stack.  And since we're using Sentry already
I searched for "distributed tracing sentry" and got to:

https://sentry.io/features/distributed-tracing/

Which shows JUST the sort of display I want to see.

**Comment by rahulbot** [comment]
Very helpful take and great potential tool to help! From a quick read it looks like we'd need to add some integration work at each touchpoint on that train of calls you laid out. I'll work with Emily on a list of 4-5 queries that capture different types of use cases we want to optimize for.

**Comment by pgulley** [comment]
I've just draft-pr'd a very trivial change to web-search to enable tracing in the first case on web-search: https://github.com/mediacloud/web-search/pull/655

[Per sentry:](https://docs.sentry.io/platforms/python/performance/instrumentation/custom-instrumentation/) we might want to think about adding custom instrumentation to the providers calls, but just getting the basic tracing enabled there would do a lot for to start

**Comment by pgulley** [comment]
Additionally, [here's](https://github.com/mediacloud/story-indexer/pull/282) another PR that addresses an issue that was preventing the sentry_sdk from getting initialized in the news-search-api deployment- at least as long as that's getting deployed as a part of the story-indexer

**Comment by rahulbot** [comment]
Closing this in favor of the sub-tasks split off onto new issues.
